File: framework/src/datamirai_engine/llm/adapters/openrouter.py
"""OpenRouter adapter — multi-provider routing via OpenAI-compatible API."""

from __future__ import annotations

from typing import Any

from datamirai_engine.llm.adapter import ModelInfo, NormalizedResponse
from datamirai_engine.llm.adapters.openai_adapter import OpenAIAdapter

_CONTEXT_WINDOWS: dict[str, int] = {
    "anthropic/claude-sonnet-4": 200000,
    "openai/gpt-4o": 128000,
    "google/gemini-2.5-flash": 1048576,
    "meta-llama/llama-3.3-70b-instruct": 131072,
}


class OpenRouterAdapter(OpenAIAdapter):
    """Adapter for OpenRouter. Inherits OpenAI format with extra headers."""

    provider_name = "openrouter"

    def __init__(
        self,
        *,
        api_key: str,
        base_url: str = "https://openrouter.ai/api",
        timeout: float = 60.0,
        site_url: str = "",
        site_name: str = "Data Mirai Engine",
    ) -> None:
        super().__init__(api_key=api_key, base_url=base_url, timeout=timeout)
        self.site_url = site_url
        self.site_name = site_name

    def _headers(self) -> dict[str, str]:
        headers = super()._headers()
        if self.site_url:
            headers["HTTP-Referer"] = self.site_url
        if self.site_name:
            headers["X-Title"] = self.site_name
        return headers

    async def call(
        self,
        *,
        model: str,
        prompt: str,
        context: str | None = None,
        temperature: float = 0.7,
        max_tokens: int = 1024,
        **kwargs: Any,
    ) -> NormalizedResponse:
        result = await super().call(
            model=model,
            prompt=prompt,
            context=context,
            temperature=temperature,
            max_tokens=max_tokens,
            **kwargs,
        )
        return NormalizedResponse(
            response=result.response,
            tokens_used=result.tokens_used,
            model=result.model,
            provider=self.provider_name,
        )

    async def list_models(self) -> list[ModelInfo]:
        models = await super().list_models()
        return [
            ModelInfo(
                id=m.id,
                name=m.name,
                context_window=_CONTEXT_WINDOWS.get(m.id) or m.context_window,
                supports_streaming=m.supports_streaming,
                supports_tools=m.supports_tools,
            )
            for m in models
        ]

    async def embed(self, text: str, *, model: str | None = None) -> list[float]:
        raise NotImplementedError(
            f"{self.provider_name} adapter does not support embeddings. "
            "Configure a provider with embedding support as fallback."
        )

    def _default_test_model(self) -> str:
        return "openai/gpt-4o-mini"

File: engine/src/llm/openrouter.rs
//! OpenRouter adapter — multi-provider routing via OpenAI-compatible API.

use std::collections::HashMap;
use std::time::Duration;

use super::openai_compat::OpenAICompatAdapter;

/// Default base URL for the OpenRouter API.
const BASE_URL: &str = "https://openrouter.ai/api/v1";

/// Default model for OpenRouter.
const DEFAULT_MODEL: &str = "meta-llama/llama-3.3-70b-instruct";

/// Build a [`OpenAICompatAdapter`] configured for OpenRouter.
///
/// OpenRouter requires two extra headers (`HTTP-Referer` and `X-Title`) for
/// attribution. These are included automatically.
pub fn new(api_key: &str) -> OpenAICompatAdapter {
    let mut extra_headers = HashMap::new();
    extra_headers.insert("HTTP-Referer".to_string(), "https://datamirai.com".to_string());
    extra_headers.insert("X-Title".to_string(), "OpenMirai".to_string());

    OpenAICompatAdapter::new(
        api_key,
        BASE_URL,
        "openrouter",
        DEFAULT_MODEL,
        extra_headers,
        Duration::from_secs(60),
    )
}

/// Build an OpenRouter adapter with custom site URL, site name, and timeout.
pub fn with_options(
    api_key: &str,
    site_url: &str,
    site_name: &str,
    timeout: Duration,
) -> OpenAICompatAdapter {
    let mut extra_headers = HashMap::new();
    if !site_url.is_empty() {
        extra_headers.insert("HTTP-Referer".to_string(), site_url.to_string());
    }
    if !site_name.is_empty() {
        extra_headers.insert("X-Title".to_string(), site_name.to_string());
    }

    OpenAICompatAdapter::new(
        api_key,
        BASE_URL,
        "openrouter",
        DEFAULT_MODEL,
        extra_headers,
        timeout,
    )
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::llm::LLMAdapter;

    #[test]
    fn openrouter_adapter_defaults() {
        let adapter = new("test-key");
        assert_eq!(adapter.provider_name(), "openrouter");
        assert_eq!(adapter.default_model(), DEFAULT_MODEL);
    }

    #[test]
    fn openrouter_custom_options() {
        let adapter = with_options(
            "test-key",
            "https://mysite.com",
            "My App",
            Duration::from_secs(120),
        );
        assert_eq!(adapter.provider_name(), "openrouter");
    }
}
